ATT-24 (work in progress)

ridmal opened the pull request.

Commit:
ATT-24 REST end point to query for lists of attachments

Commit:
ATT-24 REST end point to query for lists of attachments, modified code on first review

Commit:
ATT-24-1 Add AttachmentsService to api

Commit:
ATT-24-1 Add AttachmentsService to api, modified the code on reviews

Commit:
ATT-24-1 Add AttachmentsService to api, used @qualifier annotation following reviews

Commit:
ATT-24 modified the code in TestHelper

Commit:
ATT-24-1 modified the code according to reviews

Commit:
ATT-24-1 modified the code according to reviews 3

Commit:
ATT-24-1 modified the code to check for uuid of each element

Commit:
ATT-24-1 modified the pom.xml to add missing dependencies

Commit:
ATT-24-1 removed modifications in TestingApplicationContext.xml

Commit:
ATT-24-1 modified pom.xml

Commit:
ATT-24-1 modified test case and added test scope for dependencies

Commit:
ATT-24-1 modified test case to add non-complex obs

Commit:
ATT-24-1 modified test case to add non-complex obs

Commit:
ATT-24-1 modified the logic to get conceptComplexList

Commit:
ATT-24-1 added error logging to check for null concepts returned by conceptService

Commit:
ATT-24-1 saved a non-complex Obs in the test method.

Commit:
ATT-24-1 modified AttachmentsService.

Commit:
ATT-24-1 added test method to test for visit parameter

ridmal committed:
Merge branch 'ATT-24-1' of https://github.com/dilekamadushan/openmrs-module-attachments into ATT-24

ridmal committed:
add complex obs for test

ridmal committed:
Testing complex data with cocept

ridmal committed:
Add non-attachment complex obs for testing

chinzou commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -0,0 +1,122 @@
+		Obs obs = complexObsList.get(0);
+		Patient patient = obs.getEncounter().getPatient();
+		Encounter encounter = obs.getEncounter();
+		Visit visit = encounter.getVisit();
```
`visit` has never been used, i think you can just get rid of it.

chinzou commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -0,0 +1,122 @@
+		List<Attachment> actualAttachments = as.getAttachments(patient, encounter, true);
+		
+		// Verify
+		List<Attachment> expectedAttachments = complexObsList.stream().map(Attachment::new).collect(Collectors.toList());
```
fetching the expected attachments belongs to `// Reply` section, i think it's better to keep only Assertions in `// verify`

mks-d replied:
@ridmal @chinzou `// replay` (not `// reply`).

chinzou commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -0,0 +1,122 @@
+		Obs conceptComplexDefaultObs = testHelper.saveComplexObsForEncounterDefault(patient, conceptComplex, encounter);
+		
+		// saving some other obs during the same encounter
+		Obs otherObs = new Obs();
```
i think creating an Obs should be done in a separate scoop as it doesn't impact directly the test.

ridmal replied:
Can you explain what do you mean by "done in separate scope" ? This obs help to make sure that AttachmentService will not fetch the every obs relevant to the particular encounter and patient.

chinzou replied:
i mean enclose it in a bloc, an example is [Here](https://github.com/chinzou/openmrs-module-attachments/blob/0730834595b3bc1414a3cb93f73eaab207984a44/omod/src/test/java/org/openmrs/module/attachments/web/controller/AttachmentsControllerTest.java#L113)

ridmal committed:
merge all the Obs creations inside the one method and modify the test cases

ridmal commented:
I did the changes you mentioned @chinzou and i insert all the Obs creations into one method. Also modify the getAttachments_shouldReturnVisitAttachments() test case.

ridmal committed:
modify the getAttachments_shouldReturnVisitAttachments() method

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java:
```diff
@@ -0,0 +1,362 @@
+import liquibase.util.file.FilenameUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.RandomStringUtils;
+import org.openmrs.*;
```
Please avoir wildcard imports.
@chinzou could you direct @ridmal into configuring IntelliJ for not doing so?

chinzou replied:
@ridmal 
to avoid using wildcard imports in Intellij.
go to : File > Settings , Then Editor > Code Style > Java > Imports
you'll find `Class count to use import with '*':` property, just set it to a very high value. like 999 :p
more details [here](https://stackoverflow.com/questions/3348816/intellij-never-use-wildcard-imports)

ridmal replied:
Ok , Thank you. @chinzou and @mks-d

mks-d commented on api/src/main/java/org/openmrs/module/attachments/AttachmentsServiceImpl.java:
```diff
@@ -0,0 +1,113 @@
+	}
+	
+	@Override
+	public List<Attachment> getAttachments(Patient patient, Visit visit, boolean includeRetired) {
```
Unless there is a performance drag, why not do something a bit more comprehensive such as
- Fetching the list of encounters for the visit.
- Reusing the above `getAttachments(patient, encounter, includeRetired)` for each encounter of the visit.
- Concatenating the results.

At least like that you would _reuse_ an existing implemented routine.

mks-d commented on api/src/main/java/org/openmrs/module/attachments/AttachmentsServiceImpl.java:
```diff
@@ -0,0 +1,113 @@
+		return attachmentsList;
+	}
+	
+	public List<Attachment> getAttachments(Patient patient, boolean includeRetired) {
```
In the light of my last remark, how would we approach this one in your opinion?

mks-d reviewed (changes requested):
There is two `TestHelper` now, one in **/api** and one in **/api-2.0**. I doubt this was necessary.

ridmal committed:
remove the Additional TestHelper and modify the AttachmentServiceImpl

mks-d commented on api/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java:
```diff
@@ -62,6 +70,8 @@
 	
 	protected MockMultipartFile lastSavedMultipartImageFile;
 	
+	protected ConceptComplex ConceptComplexOutAttach;
```
`ConceptComplexOutAttach` is an invalid Java variable naming. Those cannot start with a capital letter, this is reserved for class names.
Anyway, make it `otherConceptComplex` and simply declare it as a `Concept` (not `ConceptComplex`).

mks-d commented on api/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java:
```diff
@@ -117,6 +127,24 @@ public void init() throws IOException {
 		    new GlobalProperty(AttachmentsConstants.GP_ENCOUNTER_TYPE_UUID, AttachmentsConstants.ENCOUNTER_TYPE_UUID));
+		
+		if (context.getConceptService().getConceptByName("OutOfAttachmentsTestComplex") == null) {
+			ConceptComplex conceptComplexOutAttach = new ConceptComplex();
```
Within the `if` scope you can just  name that one `conceptComplex`:
```java
ConceptComplex conceptComplex = new ConceptComplex();
```

mks-d commented on api/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java:
```diff
@@ -234,4 +262,148 @@ public String getTestComplexObsFilePath() {
+	 *
+	 * @param count The number of the complex obs to be saved.
+	 */
+	public List<Obs> saveComplexObsForEncounter(int count) throws IOException {
```
Could you get rid for now of `saveComplexObs(int)` and `saveComplexObsForVisit(int)` and focus on `saveComplexObsForEncounter` for the sake of our first test.
Change its name and signature so that the target encounter is passed as a parameter, let make it something like that:
```java
public List<Obs> saveComplexObs(Encounter encounter, int count, int otherCount) {
  ...
}
```
Where `count` is the number of attachment-complex obs to save within the encounter and `otherCount` is the number of other complex obs to save within the encounter.

When you have that one up, we can look at the first unit test.

ridmal replied:
What about normal Obs  creation for given encounter( not complex obs ). Is it not necessary for this test ?

mks-d replied:
If we want to do that (add regular obs in the mix), we can do it in the setup of our test, we don't need to encompass it in `TestHelper`'s API.

ridmal committed:
modify the saveCompelxObs method

ridmal committed:
modify the saveCompelxObs method

mks-d commented on api/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java:
```diff
@@ -198,4 +239,167 @@ public String getComplexObsDir() {
+	 *
+	 * @param count The number of the complex obs to be saved.
+	 */
+	public List<Obs> saveComplexObs(int count) throws IOException {
```
Do we still need that one?

ridmal replied:
Ok I will remove that.

mks-d commented on api/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java:
```diff
@@ -198,4 +239,167 @@ public String getComplexObsDir() {
+			// Creating the complex Obs not relevent / relevent to the attachment module
+			// during the same encounter.
+			Obs obs = new Obs();
+			if (allComplexRelatedToAttach) {
```
I don't understand this `if` statement, why don't you just loop `otherCount` times to create as many non-attachments related complex obs as specified using `otherConceptComplex` as an underlying concept complex?

That was simply the plan, did I miss something?

ridmal replied:
Initially you mentioned that we should able to configure that non-attachments to attachment related and check with the AttachmentService. 

"Edit the configuration and encompass one (or two) of the other complex obs to be also Attachments-managed.
Verify that the service returns the two original Attachments-managed ones plus the one (or two) that were encompassed through the configuration change. "

That's why I used that boolean varibale for configure that it is non-attachment or attachment related. Is It ok ? ( or is it a wrong way ?)

mks-d commented on api/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java:
```diff
@@ -198,4 +239,167 @@ public String getComplexObsDir() {
+		byte[] randomData = new byte[20];
+		Patient patient = encounter.getPatient();
+		Visit visit = encounter.getVisit();
+		for (int i = 0; i < count; i++) {
```
This loop up to `count` already adds attachments-complex obs.

mks-d commented on api/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java:
```diff
@@ -198,4 +239,167 @@ public String getComplexObsDir() {
+			obsList.add(obsSaver.saveOtherAttachment(visit, patient, encounter, fileCaption, multipartRandomFile,
+			    ValueComplex.INSTRUCTIONS_DEFAULT));
+		}
+		for (int i = 0; i < otherCount; i++) {
```
This second loop up to `otherCount` is specific to adding non-attachments complex obs. "Other" refers to complex obs that are coded with a concept that is _other than for attachments_.

Does that make sense?

ridmal replied:
OK , I understood it wrongly. Sorry. I will quickly changed it and update the PR.

ridmal committed:
Changs in TestHelper saveComplexObs method

mks-d reviewed (commented):
@ridmal I left minor comments directly on your last commit. Looking good!

ridmal commented:
@mksd , thank you and i will add the changes.

ridmal committed:
Add method for test the fetch attachment based on visits

mks-d commented on api/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java:
```diff
@@ -114,6 +128,39 @@ public void init() throws IOException {
 		context.getAdministrationService().saveGlobalProperty(
 		    new GlobalProperty(AttachmentsConstants.GP_ENCOUNTER_TYPE_UUID, AttachmentsConstants.ENCOUNTER_TYPE_UUID));
+		
+		// Create a out of attachment complex concept.
```
Change this comment to
```java
// Create a concept complex that is not managed by Attachments
```

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -0,0 +1,114 @@
+		List<Attachment> expectedAttachments = testHelper.saveComplexObs(encounter, 2, 2);
+		Patient patient = encounter.getPatient();
+		
+		{
```
You can remove that scope. It's great that you go to the length of adding even a non-complex obs, however you haven't done that for the 'visit' case.

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -0,0 +1,114 @@
+		Patient patient = encounter.getPatient();
+		
+		{
+			// Createing some other obs ( not complex Obs ) during the same encounter
```
Typo, anyway just set a simple comment such as:
```java
// adding a non-complex obs
```

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -0,0 +1,114 @@
+	@Test
+	public void getAttachments_shouldReturnVisitAttachments() throws Exception {
+		
+		Encounter encounter = testHelper.getTestEncounter();
```
For consistency name this as `encounter1`.

mks-d replied:
Prefix this line with the following comment:
```java
// setup
```

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -0,0 +1,114 @@
+		expectedAttachments.addAll(testHelper.saveComplexObs(encounter2, 2, 2));
+		expectedAttachments.addAll(testHelper.saveComplexObs(encounter3, 3, 3));
+		
+		List<Attachment> actualAttachments = as.getAttachments(patient, visit, true);
```
Prefix this line with the following comment:
```java
// replay
```

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -0,0 +1,114 @@
+		
+		List<Attachment> actualAttachments = as.getAttachments(patient, visit, true);
+		
+		Assert.assertArrayEquals(
```
Prefix this line with the following comment:
```java
// verif
```

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -0,0 +1,114 @@
+	@Test
+	public void getAttachments_shouldReturnEncounterAttachments() throws Exception {
+		
+		Encounter encounter = testHelper.getTestEncounter();
```
Prefix this line with the following comment:
```java
// setup
```

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -0,0 +1,114 @@
+			otherObs = os.saveObs(otherObs, null);
+		}
+		
+		List<Attachment> actualAttachments = as.getAttachments(patient, encounter, true);
```
Prefix this line with the following comment:
```java
// replay
```

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -0,0 +1,114 @@
+		
+		List<Attachment> actualAttachments = as.getAttachments(patient, encounter, true);
+		
+		Assert.assertArrayEquals(
```
Prefix this line with the following comment:
```java
// verif
```

mks-d reviewed (commented):
Very good @ridmal . I have some minor comments, I will follow up on Talk for the rest.

I encourage you to highlight each test case's subsections following the structure **setup** → **replay** → **verif**.

This clarifies unit tests a lot, and forces the developer to highlight the actual replay (which is what is actually tested).

mks-d commented on api/src/main/java/org/openmrs/module/attachments/AttachmentsServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+	public List<Attachment> getAttachments(Patient patient, Encounter encounter, boolean includeRetired) {
+		
+		List<Person> personList = new ArrayList<>();
+		List<Attachment> attachmentsList = new ArrayList<>();
```
You don't need the `List` suffix, just name collections as a plural of what they hold.

- `attachmentsList` → `attachments`
- `personList ` → `persons`
- `encounterList ` → `encounters`
- ...

mks-d commented on api/src/main/java/org/openmrs/module/attachments/AttachmentsServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+			} else {
+				questionConcepts.add(concept);
+			}
+		}
```
At this point if the list of question concepts is empty what happens? I would assume that an `APIException` should be thrown. You could set a message like:
>No concepts complex were configured or found to query attachments.

ridmal committed:
Add getIsolatedAttachments method in Attachment API

ridmal committed:
Add Javadoc for the AtachmentService

mks-d commented on api/src/main/java/org/openmrs/module/attachments/AttachmentsService.java:
```diff
@@ -11,13 +12,48 @@
+
+
+	/**
+	 * Get all attachments for given patient.
```
→ `Get a patient's attachments that are associated with encounters.`

mks-d commented on api/src/main/java/org/openmrs/module/attachments/AttachmentsService.java:
```diff
@@ -11,13 +12,48 @@
+	/**
+	 * Get all attachments for given patient.
+	 *
+	 * @param patient the patient whose attachment to get
```
You can remove that line, the parameter is obvious in the context of OpenMRS.

mks-d commented on api/src/main/java/org/openmrs/module/attachments/AttachmentsService.java:
```diff
@@ -11,13 +12,48 @@
+	 * Get all attachments for given patient.
+	 *
+	 * @param patient the patient whose attachment to get
+	 * @param includeRetired
```
→ `Specifies whether the underlying complex obs that are fetched should include retired ones or not.`

mks-d commented on api/src/main/java/org/openmrs/module/attachments/AttachmentsService.java:
```diff
@@ -11,13 +12,48 @@
+	 * @param patient the patient whose attachment to get
+	 * @param includeRetired
+	 * @return matching attachments
+	 * @throws APIException
```
You should say _when_ an API exception is thrown.

mks-d commented on api/src/main/java/org/openmrs/module/attachments/AttachmentsService.java:
```diff
@@ -11,13 +12,48 @@
+	 *
+	 * @param patient the patient whose attachment to get
+	 * @param includeRetired
+	 * @return matching attachments
```
You can remove that line, the returned result is obvious.

mks-d commented on api/src/main/java/org/openmrs/module/attachments/AttachmentsService.java:
```diff
@@ -11,13 +12,48 @@
-	
+
+	/**
+	 * Get all attachments for given encounter.
```
→ `Get a patient's attachments that are associated with a specified encounter.`

mks-d commented on api/src/main/java/org/openmrs/module/attachments/AttachmentsService.java:
```diff
@@ -11,13 +12,48 @@
-	
+
+	/**
+	 * Get all attachments for given visit.
```
→ `Get a patient's attachments that are associated with a specified visit.`

mks-d commented on api/src/main/java/org/openmrs/module/attachments/AttachmentsService.java:
```diff
@@ -11,13 +12,48 @@
-	
+
+	/**
+	 * Get all attachments for given patient which are not associated with any encounters.
```
→ `Get a patient's attachments that are not associated with any visits or encounters.`

mks-d reviewed (changes requested):
I have detailed changes to make to the first Javadoc in details, please apply those changes to the others as well. I have also reworded the headline English.

Thos methods are quite obvious in the context of OpenMRS, so in the end the Javadoc can stay quite concise and simple.

mks-d reviewed (changes requested):
We need at least one unit test case for the new `getIsolatedAttachments(..)`.

ridmal committed:
add unit test for getIsolatedAttachments() method

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -0,0 +1,187 @@
+		}
+		
+		// Saves a complex obs as if they had been saved outside to the attachments not associated with any visits or encounters.
+		for (int j = 0; j < 2; j++) {
```
Reuse `i` as a name for the counter variable. Those loop counters' scopes are _inside_ the `for` statement, you can have as many successive `for` statements that keep reusing `i` as the counter variable name.

mks-d replied:
If you have done this in other places, please apply the same logic there as well.

mks-d reviewed (changes requested):
Reuse `ComplexObsSaver`! ;-)
Passing a `null` visit and encounter will achieve the saving of an 'isolated' attachment. See my modifications below:

```java
@Autowired
@Qualifier(AttachmentsConstants.COMPONENT_COMPLEXOBS_SAVER)
private ComplexObsSaver saver;

...

@Test
public void getAttachments_shouldReturnIsolatedAttachments() throws Exception {
  
  // setup
  Encounter encounter = testHelper.getTestEncounter();
  testHelper.saveComplexObs(encounter, 3, 2);
  Patient patient = ps.getPatient(2);
  
  // attachments not bound to any visits/encounters
  List<Attachment> expectedAttachments = new ArrayList<>();
  for (int i = 0; i < 2; i++) {
    byte[] randomData = new byte[20];
    new Random().nextBytes(randomData);
    
    MockMultipartFile multipartRandomFile = new MockMultipartFile("1", "1", "application/octet-stream", randomData);
    Obs obs = saver.saveOtherAttachment(null, patient, null, "File caption #" + i + 1, multipartRandomFile,
        ValueComplex.INSTRUCTIONS_DEFAULT);
    expectedAttachments.add(new Attachment(obs));
  }
  
  // adding some non-complex obs
  for (int i = 0; i < 2; i++) {
    Obs otherObs = new Obs();
    otherObs.setConcept(cs.getConcept(3));
    otherObs.setObsDatetime(new Date());
    otherObs.setPerson(patient);
    otherObs.setValueText("Some text value for a test obs # " + i + 1);
    otherObs = os.saveObs(otherObs, null);
  }
  
  // replay
  List<Attachment> actualAttachments = as.getIsolatedAttachments(patient, true);
  
  // verify
  Assert.assertArrayEquals(
      expectedAttachments.stream().map(Attachment::getUuid).collect(Collectors.toList()).toArray(),
      actualAttachments.stream().map(Attachment::getUuid).collect(Collectors.toList()).toArray());
}
```

ridmal committed:
modify the getAttachments_shouldReturnIsolatedAttachments() method

ridmal committed:
changes in Attachment API

mks-d reviewed (changes requested):
What happened to `getIsolatedAttachments(..)`?

Could your re-add that method and its tests and also add one more test that verifies that
>attachments with `includeIsolated=true` = attachments with `includeIsolated=false` + isolated attachments

ridmal commented:
Oh sorry , I thought that getIsolatedAttachments(..) method is not necessary now. I will add that again.

ridmal committed:
add unit test for  IncludeIsolated parameter

mks-d committed:
ATT-24: Streamlining tests setup + rearranging APIs.

  modified:   api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java
  modified:   api/src/main/java/org/openmrs/module/attachments/AttachmentsService.java
  modified:   api/src/main/java/org/openmrs/module/attachments/AttachmentsServiceImpl.java
  modified:   api/src/test/java/org/openmrs/module/attachments/obs/ImageAttachmentHandlerContextSensitiveTest.java
  modified:   api/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java
  modified:   omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentController1_10Test.java
  modified:   omod-2.0/src/test/java/org/openmrs/module/attachments/rest/AttachmentController2_0Test.java

ridmal committed:
add unit test for check Attachment API will thrown APIException for misconfigured complex obs

mks-d commented on api/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java:
```diff
@@ -114,9 +114,9 @@ public void init() throws IOException {
-		                "[\"7cac8397-53cd-4f00-a6fe-028e8d743f8e\",\"42ed45fd-f3f6-44b6-bfc2-8bde1bb41e00\"]"));
+		context.getAdministrationService().saveGlobalProperty(new GlobalProperty(
+		        AttachmentsConstants.GP_CONCEPT_COMPLEX_UUID_LIST,
+		        "[\"7cac8397-53cd-4f00-a6fe-028e8d743f8e\",\"42ed45fd-f3f6-44b6-bfc2-8bde1bb41e00\",\"32d3611a-6699-4d52-823f-b4b788bac3e3\"]"));
```
This specific setup should only belong to the one test that aserts the `APIException`, we don't want the misconfiguration to affect all tests.

mks-d replied:
@ridmal the above comment has not been addressed yet, could you look into it?

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -238,4 +245,66 @@ public void getAttachments_shouldReturnIsolatedAttachments() throws Exception {
+	}
+	
+	@Test(expected = APIException.class)
+	public void getAttachments_shouldThrowAPIExceptionForMisconfiguredComplexObs() throws Exception {
```
→ `getAttachments_shouldThrowWhenFetchingNonComplexObs`

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -238,4 +245,66 @@ public void getAttachments_shouldReturnIsolatedAttachments() throws Exception {
 	}
+	
+	@Test
+	public void testIncludeIsolatedParameter() throws Exception {
```
Respect the test routines naming convention.
→ `getAttachments_shouldBeConsistentAcrossServiceApi`

mks-d committed:
ATT-24: Setting Attachment's obs date-time and all BaseOpenmrsData.

  modified:   api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java
  modified:   api/src/main/java/org/openmrs/module/attachments/obs/Attachment.java

ridmal committed:
changed term isolated to encounterless

mks-d commented on api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java:
```diff
@@ -144,7 +147,7 @@ public void getAttachments_shouldReturnAllAttachments() throws Exception {
 		// attachments not bound to any visits/encounters
 		expectedAttachments.addAll(asAttachments(testHelper.saveComplexObs(null, 2, 0)));
-		
+		Collections.sort(expectedAttachments, Comparator.comparing(Attachment::getDateTime).reversed());
```
Nice! This this solve the issue then?

ridmal replied:
Yes , It solved the issue . Others test cases will not give this error in my machine. But i think that adding this sorting method to all the other test cases will make sure that this issue is not occur again for any reason.

mks-d committed:
ATT-24: Streamlining the sorting of attachments in unit tests.

  modified:   api-2.0/src/test/java/org/openmrs/module/attachments/AttachmentsServiceTest.java
  modified:   api/src/test/java/org/openmrs/module/attachments/obs/TestHelper.java

ridmal committed:
Modify the getAttachments_shouldThrowWhenFetchingNonComplexObs test case

mks-d committed:
ATT-24: Clarifying the setup of test about module misconfiguration.

mks-d commented:
Big thanks @ridmal for all your work on this PR. There is enough now to commit this and move on to the REST API.

@dilekamadushan thanks for your early work on this that could become a starting point for @ridmal .
